### ISSUE TITLE
F.mse_loss(a, b, reduction='elementwise_mean') value is incorrect and doesn't show deprecation warning when 2nd argument requires gradient

### ISSUE THREAD

**Issue by vitchyr** [pytorch/pytorch#43228]
## 🐛 Bug

F.mse_loss(a, b, reduction='elementwise_mean') has very different behaviors depending on if `b` require a gradient or not.

## To Reproduce

```
import torch
from torch.nn import functional as F

A = torch.ones(2)
B = torch.zeros(2, requires_grad=True)
print(F.mse_loss(A, B, reduction='elementwise_mean'))

C = torch.zeros(2)
print(F.mse_loss(A, C, reduction='elementwise_mean'))
```

returns
```
# first call takes the sum
tensor(2., grad_fn=<SumBackward0>)

# second call takes the mean and prints a warning
/home/vitchyr/anaconda2/envs/tmp/lib/python3.6/site-packages/torch/nn/_reduction.py:14: UserWarning: reduction='elementwise_mean' is deprecated, please use reduction='mean' instead.
  warnings.warn("reduction='elementwise_mean' is deprecated, please use reduction='mean' instead.")
tensor(1.)
```

## Expected behavior

I think the second behavior (a deprecation warning and taking the mean) is the correct/expected behavior, but the two behaviors do not match.

## Environment
```
PyTorch version: 1.6.0
Is debug build: False
CUDA used to build PyTorch: 10.2

OS: Ubuntu 16.04.6 LTS (x86_64)
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.12) 5.4.0 20160609
Clang version: Could not collect
CMake version: version 3.5.1

Python version: 3.6 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: 10.1.105
GPU models and configuration: 
GPU 0: GeForce GTX 1080
GPU 1: GeForce GTX 1080

Nvidia driver version: 418.56
cuDNN version: Probably one of the following:
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.6
/usr/local/cuda-10.1/targets/x86_64-linux/lib/libcudnn.so.7

Versions of relevant libraries:
[pip] numpy==1.16.3
[pip] torch==1.6.0
[conda] numpy                     1.16.3                   pypi_0    pypi
[conda] torch                     1.6.0                    pypi_0    pypi
```
## Additional context

cc @ezyang @gchanan @zou3519 @yf225 @glaringlee @albanD @mruberry

**Comment by SsnL** [comment]
This looks pretty bad. Tentatively assign hi-pri. The triage folks can discuss and decide.

**Comment by gchanan** [comment]
@kshitij12345 patch probably fixes the exact issue reported here, but there are a number of other problems, namely:
1) there is a hacky autograd path that doesn't even validate the reduction string is valid
2) the hacky autograd path doesn't support all the reduction types
3) the hacky autograd path is slow
4) the hacky autograd path is not tested

So, I'll put up a bigger fix.

**Comment by gchanan** [comment]
Also the C++ API has the same issue.

**Comment by gchanan** [comment]
This has been fixed by https://github.com/pytorch/pytorch/pull/44437 and other PRs in the stack fixed similar issues for other loss functions.
